dask/dot.py

```diff
@@ -132,19 +133,27 @@ def to_graphviz(dsk, data_attributes=None, function_attributes=None):
 
 def dot_graph(dsk, filename='mydask', **kwargs):
     g = to_graphviz(dsk, **kwargs)
-    g.save(filename + '.dot')
 
-    try:
-        check_call('dot -Tpdf {0}.dot -o {0}.pdf'.format(filename), shell=True)
-        check_call('dot -Tpng {0}.dot -o {0}.png'.format(filename), shell=True)
-    except CalledProcessError:
-        raise RuntimeError(
-            "Please install The `dot` utility from graphviz:\n"
-            "  Debian:  sudo apt-get install graphviz\n"
-            "  Mac OSX: brew install graphviz\n"
-            "  Windows: http://www.graphviz.org/Download..php")  # pragma: no cover
+    if filename is not None:
+        g.save(filename + '.dot')
+
+        try:
+            check_call('dot -Tpdf {0}.dot -o {0}.pdf'.format(filename),
+                       shell=True)
+            check_call('dot -Tpng {0}.dot -o {0}.png'.format(filename),
+                       shell=True)
+
+        except CalledProcessError:
+            msg = ("Please install The `dot` utility from graphviz:\n"
+                   "  Debian:  sudo apt-get install graphviz\n"
+                   "  Mac OSX: brew install graphviz\n"
+                   "  Windows: http://www.graphviz.org/Download..php")
+            raise RuntimeError(msg)  # pragma: no cover
     try:
         from IPython.display import Image
-        return Image(filename + '.png')
+        s = BytesIO()
+        s.write(g.pipe(format='png'))
```

Comment:
Should we use this rather than `check_call` above?  

If a filename is provided then we compute the layout one time more than is necessary.  The layout *can* become quite expensive on some larger graphs.  It may be good to reuse the computation if possible.